netsync/manager_test.go

```diff
@@ -0,0 +1,776 @@
+			!call.newTxs[0].Tx.Hash().IsEqual(tx2.Hash()) ||
+			!call.newTxs[1].Tx.Hash().IsEqual(tx3.Hash()) {
+
+			t.Fatalf("PeerNotifier received unexpected "+
+				"AnnounceNewTransactions call: %v", call.newTxs)
+		}
+	default:
+		t.Fatalf("Expected SyncManager to make " +
+			"AnnounceNewTransactions call to PeerNotifier")
+	}
+
+	// Send non-standard, invalid transaction with corrupt scriptSig
+	badScriptSig := append(scriptSig, 0xaa, 0xbb, 0xcc, 0xdd)
+	tx4, err := createSpendingTx(tx3, 0, badScriptSig, address)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Attempt to process orphan transaction
+	syncMgr.QueueTx(tx4, localNode, syncChan)
+	select {
+	case <-syncChan:
+	case <-time.After(time.Second):
+		t.Fatal("Timeout waiting for sync manager to process transaction 4")
+	}
+
+	// AnnounceNewTransactions should not be called for invalid transaction
+	select {
+	case call := <-ctx.peerNotifier.announceNewTransactionsChan:
+		t.Fatalf("PeerNotifier received unexpected "+
+			"AnnounceNewTransactions call: %v", call.newTxs)
+	default:
+	}
+
+	// Expect node to send reject in response to invalid transaction
+	select {
+	case msg := <-remoteMessages.rejectChan:
+		if msg.Code != wire.RejectNonstandard {
+			t.Fatalf("Reject message has unexpected code %s, expected %s",
+				msg.Code, wire.RejectNonstandard)
+		}
+	case <-time.After(time.Second):
+		t.Fatal("Timeout waiting for remote node to receive reject message")
+	}
+
+	// An already rejected transaction should not get a reject response
+	syncMgr.QueueTx(tx4, localNode, syncChan)
+	select {
+	case <-syncChan:
+	case <-time.After(time.Second):
+		t.Fatal("Timeout waiting for sync manager to process transaction 4")
+	}
+
+	// Expect node to send reject in response to invalid transaction
+	select {
+	case msg := <-remoteMessages.rejectChan:
+		t.Fatalf("Received unexpected reject message %v", msg)
+	case <-time.After(time.Second):
+	}
+
+	err = syncMgr.Stop()
+	if err != nil {
+		t.Fatal("failed to stop SyncManager: %v", err)
+	}
+}
+
+type msgChans struct {
+	memPoolChan    chan *wire.MsgMemPool
+	txChan         chan *wire.MsgTx
+	blockChan      chan *wire.MsgBlock
+	invChan        chan *wire.MsgInv
+	headersChan    chan *wire.MsgHeaders
+	getDataChan    chan *wire.MsgGetData
+	getBlocksChan  chan *wire.MsgGetBlocks
+	getHeadersChan chan *wire.MsgGetHeaders
+	rejectChan     chan *wire.MsgReject
+}
+
+func newMessageChans() *msgChans {
+	var instance msgChans
+	instance.memPoolChan = make(chan *wire.MsgMemPool)
+	instance.txChan = make(chan *wire.MsgTx)
+	instance.blockChan = make(chan *wire.MsgBlock)
+	instance.invChan = make(chan *wire.MsgInv)
+	instance.headersChan = make(chan *wire.MsgHeaders)
+	instance.getDataChan = make(chan *wire.MsgGetData)
+	instance.getBlocksChan = make(chan *wire.MsgGetBlocks)
+	instance.getHeadersChan = make(chan *wire.MsgGetHeaders)
+	instance.rejectChan = make(chan *wire.MsgReject)
+	return &instance
+}
+
+func buildBlockInv(block *wire.MsgBlock) *wire.MsgInv {
```

Comment:
I think this would be more useful as `blocks ...*btcutil.Block`, so you could use it for the multi-block calls too.

Something like:

```Go
func buildBlockInv(blocks ...*btcutil.Block) *wire.MsgInv {
	msg := wire.NewMsgInv()
	for _, block := range blocks {
		invVect := wire.NewInvVect(wire.InvTypeBlock, block.Hash())
		msg.AddInvVect(invVect)
	}
	return msg
}
```

Then above:

```Go
...
	// Remote node replies to getblocks with an inv.
	syncMgr.QueueInv(buildBlockInv(blocks...), localNode)
...
	// SyncManager should send a getdata message requesting blocks.
	syncMgr.QueueInv(buildBlockInv(block), localNode)
...
	// Send invalid block with timestamp in the far future.
	...
	syncMgr.QueueInv(buildBlockInv(block), localNode)
```